kroxylicious-additional-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilter.java

```diff
@@ -0,0 +1,228 @@
+        }
+        return newPartitionData;
+    }
+
+    @Override
+    public CompletionStage<ResponseFilterResult> onFetchResponse(short apiVersion, ResponseHeaderData header, FetchResponseData response, FilterContext context) {
+        return maybeDecodeFetch(response.responses())
+                .thenCompose(responses -> context.forwardResponse(header, new FetchResponseData().setResponses(responses)));
+    }
+
+    private CompletableFuture<List<FetchableTopicResponse>> maybeDecodeFetch(List<FetchableTopicResponse> topics) {
+        List<CompletableFuture<FetchableTopicResponse>> result = new ArrayList<>();
+        for (FetchableTopicResponse topicData : topics) {
+            result.add(maybeDecodePartitions(topicData.partitions()).thenApply(kk -> {
+                topicData.setPartitions(kk);
+                return topicData;
+            }));
+        }
+        return join(result);
+    }
+
+    private CompletableFuture<List<PartitionData>> maybeDecodePartitions(List<PartitionData> partitions) {
+        List<CompletableFuture<PartitionData>> result = new ArrayList<>();
+        for (PartitionData partitionData : partitions) {
+            if (!(partitionData.records() instanceof MemoryRecords)) {
+                throw new IllegalStateException();
+            }
+            result.add(maybeDecodeRecords(partitionData, (MemoryRecords) partitionData.records()));
+        }
+        return join(result);
+    }
+
+    private CompletableFuture<PartitionData> maybeDecodeRecords(PartitionData fpr, MemoryRecords memoryRecords) {
+        final CompletableFuture<PartitionData> result;
+        if (!anyRecordSmellsOfEncryption(memoryRecords)) {
+            result = CompletableFuture.completedFuture(fpr);
+        }
+        else {
+            ByteBuffer buffer = null; // TODO
+            MemoryRecordsBuilder builder = recordsBuilder(buffer, memoryRecords);
+            List<CompletableFuture<Integer>> futures = new ArrayList<>();
+            for (var kafkaRecord : memoryRecords.records()) {
+                if (!smellsOfEncryption(kafkaRecord.value())) {
+                    builder.append(kafkaRecord.timestamp(), kafkaRecord.key(), kafkaRecord.value(), kafkaRecord.headers());
+                    futures.add(CompletableFuture.completedFuture(1));
+                }
+                else {
+                    ByteBuffer value = kafkaRecord.value();
+                    var cf = dekCache.resolve(value).thenApply((DekContext<K> dekContext) -> {
+
+                        ByteBuffer output = null;
+                        try {
+                            int plaintextSize = kafkaRecord.valueSize();
+                            // TODO ^^ this is an overestimate!
+                            output = bufferPool.acquire(plaintextSize);
+                            dekContext.decode(value, output);
+                            builder.append(kafkaRecord.timestamp(), kafkaRecord.key(), output, kafkaRecord.headers());
```

Comment:
It's not completely clear who responsible for the BB.flip() in your model.   AesGcmEncryptor/DekContext or this class.  The Javadoc should set out the contract in terms of the ByteBuffer's state after the call.